lib/datadog/profiling/crashtracker.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+require 'libdatadog'
+
+module Datadog
+  module Profiling
+    # Used to report Ruby VM crashes.
+    # The interesting bits are implemented as native code and using libdatadog.
+    #
+    # Methods prefixed with _native_ are implemented in `crashtracker.c`
+    class Crashtracker
+      private
+
+      attr_reader :exporter_configuration, :tags_as_array, :path_to_crashtracking_receiver_binary
+
+      public
+
+      def initialize(
+        exporter_configuration:,
+        tags:,
+        path_to_crashtracking_receiver_binary: Libdatadog.path_to_crashtracking_receiver_binary
+      )
+        @exporter_configuration = exporter_configuration
+        @tags_as_array = tags.to_a
+        @path_to_crashtracking_receiver_binary = path_to_crashtracking_receiver_binary
+      end
+
+      def start
+        start_or_update_on_fork(action: :start)
+      end
+
+      def reset_after_fork
+        start_or_update_on_fork(action: :update_on_fork)
+      end
+
+      def stop
+        begin
+          self.class._native_stop
+          Datadog.logger.debug('Crash tracking stopped successfully')
+        rescue => e
+          Datadog.logger.error("Failed to stop crash tracking: #{e.message}")
+        end
+      end
+
+      private
+
+      def start_or_update_on_fork(action:)
+        unless path_to_crashtracking_receiver_binary
+          Datadog.logger.warn(
+            "Cannot #{action} profiling crash tracking as no path_to_crashtracking_receiver_binary was found"
+          )
+          return
+        end
+
+        begin
+          self.class._native_start_or_update_on_fork(
+            action: action,
+            exporter_configuration: exporter_configuration,
+            path_to_crashtracking_receiver_binary: path_to_crashtracking_receiver_binary,
+            tags_as_array: tags_as_array,
+          )
```

Comment:
Wouldn't it be easier to have `_native_start_or_update_on_fork` be an actual method taking only a single parameter and using the private attr_readers for the rest?

Related minor: There's also currently an implicit mismatch between the native implementation which is essentially forced to work as a singleton given the crashtracker APIs and the ruby one which will happily support multiple instances if so desired but with repeated starts leading to either crashtracker update errors or silent overrides (haven't tested which happens). I think this may now become the first instance of something that would prevent us from having (if we so desired) multiple running profiler instances simultaneously? This may in practice not be that important, especially because crash tracking failures are currently just logging and not actually propagating exceptions but I feel there's a mismatch between lifecycles here.

Maybe worth "singletonning" the profiler to ensure you can never have 2 running instances of it and give a more understandable error rather than a "failed to register crashtracker with different settings" one that makes the underlying issue slightly more obscure?

Reply:
> Wouldn't it be easier to have _native_start_or_update_on_fork be an actual method taking only a single parameter and using the private attr_readers for the rest? Or does this have some unexpected consequences when dealing with native extensions?

It's definitely _possible_. I'm somewhat on the wall on it being easier. Specifically, I think it's a bit harder to debug if we're going across from Ruby to native and then back.

So I've been kinda trying to keep this pattern of, when we need to go into native, the Ruby code gathers all things that we need, and hands them over as a bunch of simple types (strings, arrays, hashes) to the native function.

But in practice in many cases (and I think this one as well), the difference isn't thaaat huge so while I slightly favor this approach, I think it'd be fine if we changed to your suggested approach (which would be a lot more object-oriented).

> Related minor: There's also currently an implicit mismatch between the native implementation which is essentially forced to work as a singleton given the crashtracker APIs and the ruby one which will happily support multiple instances if so desired but with repeated starts leading to either crashtracker update errors or silent overrides (haven't tested which happens). I think this may now become the first instance of something that would prevent us from having (if we so desired) multiple running profiler instances simultaneously? This may in practice not be that important, especially because crash tracking failures are currently just logging and not actually propagating exceptions but I feel there's a mismatch between lifecycles here.
>
> Maybe worth "singletonning" the profiler to ensure you can never have 2 running instances of it and give a more understandable error rather than a "failed to register crashtracker with different settings" one that makes the underlying issue slightly more obscure?

So this is an interesting point.

We actually already have a check that makes the profiler a singleton. Specifically the `CpuAndWallTimeWorker` has some global state in native code (`active_sampler_instance` / `active_sampler_instance_state`) and if you do try to start more than one you'll get

> Could not start CpuAndWallTimeWorker: There's already another instance of CpuAndWallTimeWorker active in a different thread

In practice, you'd only hit this if you were directly invoking `Datadog::Profiling::Component.build_profiler_component` or even creating any of the other classes by themselves. None of this is supposed to be how the profiler gets used in-production: the profiler is supposed to be created via the usual ddtrace components, and those only keep one active instance of the profiler around at any one time.

So while yes, there's ways in which multiple `Crashtracker` instances could be active, in practice this should never happen in user applications, unless the profiler is being used piecemeal, which is not really supported.

The crashtracker itself doesn't completely mind `start` being called multiple times, it's a "last writer wins" for the configuration. And stop only works once, since it stops the only one active:

```
$ bundle exec pry
[1] pry(main)> require 'ddtrace'
[3] pry(main)> ct1 = Datadog::Profiling::Crashtracker.new(exporter_configuration: [:agent, 'http://localhost:6006'], tags: {})
=> #<Datadog::Profiling::Crashtracker:0x0000633690cc9be0
 @exporter_configuration=[:agent, "http://localhost:6006"],
 @path_to_crashtracking_receiver_binary="libdatadog-crashtracking-receiver",
 @tags_as_array=[]>
[4] pry(main)> ct2 = Datadog::Profiling::Crashtracker.new(exporter_configuration: [:agent, 'http://localhost:6006'], tags: {})
=> #<Datadog::Profiling::Crashtracker:0x0000633690d94890
 @exporter_configuration=[:agent, "http://localhost:6006"],
 @path_to_crashtracking_receiver_binary="libdatadog-crashtracking-receiver",
 @tags_as_array=[]>
[5] pry(main)> ct1.start
=> true
[6] pry(main)> ct1.start
=> true
[7] pry(main)> ct2.start
=> true
[8] pry(main)> ct2.start
=> true
[9] pry(main)> ct1.stop
=> true
[10] pry(main)> ct1.stop
E, [2024-04-04T09:46:42.795319 #116895] ERROR -- ddtrace: [ddtrace] (crashtracker.rb:41:in `rescue in stop') Failed to stop crash tracking: Failed to stop the crash tracker: ddog_prof_Crashtracker_shutdown failed: No crashtracking previous signal handlers
=> true
[12] pry(main)> ct2.stop
E, [2024-04-04T09:46:48.329567 #116895] ERROR -- ddtrace: [ddtrace] (crashtracker.rb:41:in `rescue in stop') Failed to stop crash tracking: Failed to stop the crash tracker: ddog_prof_Crashtracker_shutdown failed: No crashtracking previous signal handlers
=> true
```

I did, on an earlier version of this PR, use a singleton, but found that it was a bit more awkward to enable and test. I've pushed bd0537e7c0d96caa3cbac2d0137c375e7c9c039c to more clearly document what's up here; if you're still not convinced, let me know, and I can take a stab at turning it again into a singleton and see how that looks.